docs/configs-micro-tests/system.py

```diff
@@ -0,0 +1,189 @@
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+# Authors: Jason Lowe-Power
+
+"""
+Main system for all SE experiments.
+
+This contains the default caches and memory that can be used by different
+CPU models.
+
+The system does not have a CPU model, this must be specified by a subclass.
+"""
+
+import m5
+from m5.objects import *
+from optparse import OptionParser
+
+class L1Cache(Cache):
+    """Simple L1 Cache with default values"""
+
+    assoc = 8
+    tag_latency = 1
+    data_latency = 1
+    response_latency = 1
+    mshrs = 16
+    tgts_per_mshr = 20
+    def __init__(self):
+        super(L1Cache, self).__init__()
+        pass
+
+    def connectBus(self, bus):
+        """Connect this cache to a memory-side bus"""
+        self.mem_side = bus.slave
+
+    def connectCPU(self, cpu):
+        """Connect this cache's port to a CPU-side port
+           This must be defined in a subclass"""
+        raise NotImplementedError
+
+class L1ICache(L1Cache):
+    """Simple L1 instruction cache with default values"""
+
+    # Set the default size
+    size = '32kB'
+
+    def connectCPU(self, cpu):
+        self.cpu_side = cpu.icache_port
+        """Connect this cache's port to a CPU icache port"""
+    def __init__(self):
+        super(L1ICache, self).__init__()
+
+
+class L1DCache(L1Cache):
+    """Simple L1 data cache with default values"""
+
+    # Set the default size
+    size = '32kB'
+
+    def connectCPU(self, cpu):
+        """Connect this cache's port to a CPU dcache port"""
+        self.cpu_side = cpu.dcache_port
+    def __init__(self):
+        super(L1DCache, self).__init__()
+
+class L2Cache(Cache):
+    """Simple L2 Cache with default values"""
+
+    # Default parameters
+    size = '1MB'
+    assoc = 16
+    tag_latency = 10
+    data_latency = 10
+    response_latency = 1
+    mshrs = 20
+    tgts_per_mshr = 12
+
+    def connectCPUSideBus(self, bus):
+        self.cpu_side = bus.master
+
+    def connectMemSideBus(self, bus):
+        self.mem_side = bus.slave
+    def __init__(self):
+        super(L2Cache, self).__init__()
+
+class BaseTestSystem(System):
+    """Base class for all test systems.
+
+    Each test system must set its own CPU model by setting the CPUModel
+    attribute.
+    """
+    _CPUModel = BaseCPU
+    _MemoryModel = SimpleMemory
+    def __init__(self):
+
+        super(BaseTestSystem,self).__init__()
+        if self._MemoryModel is None:
```

Comment:
So, if `_MemoryModel` is None, then you have caches? That seems strange.

Do you expect users to want caches or not? Which should be default?

Reply:
Updated this so that you will have caches only with Slow Memory and not  with other memories (Inf, SingleCycle) where I thought it does not make a lot of sense to have caches